### PR TITLE
Encode some datums in a variable-length way

### PR DESCRIPTION
**Pull request by umanwizard**
Slightly belated part 2 of https://github.com/MaterializeInc/materialize/pull/4646

This adds some new tag variants for timestamps as well as signed ints, allowing us to encode them in less space in most cases. For timestamps, we only use the full 16-byte representation if we can't fit in 8 bytes. For integers, we write down only the amount of bytes we actually need.

There are two things we could tweak to get even more savings:

1. We could do this for unsigned integers as well. I didn't bother because AFAIK they are rare enough that it wasn't worth consuming a bunch more tag space.
2. Currently we still use one bit to encode the sign, which roughly halves the set of values representable at each level.  We could encode it in the tag instead (doubling the amount of new tags).

### CODE REVIEW

File: src/repr/src/row.rs [hunk 1]
```diff
@@ -746,16 +874,58 @@ where
...
+            } else if mbs <= 48 {
+                data.push(Tag::Int64_48.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..6]);
+            } else if mbs <= 56 {
+                data.push(Tag::Int64_56.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..7]);
+            } else {
+                data.push(Tag::Int64.into());
+                data.extend_from_slice(&i.to_le_bytes());
+            }
```
Comment (guswynn) [inline src/repr/src/row.rs:928]: can we just use https://docs.rs/integer-encoding/latest/integer_encoding/trait.VarInt.html?

Comment (umanwizard) [reply src/repr/src/row.rs:928]: I think that will be less efficient than what we're doing here, which we're able to do because the tag gives us room to encode the length for free.

Comment (guswynn) [reply src/repr/src/row.rs:928]: ah yes, nice!

**Comment by benesch** [comment]
First of all: super exciting to see work towards easy wins on memory usage.

Won't have time to review the code, but did want to drop in with a rollout question! How risky is this change? It seems there is some small risk that the increased CPU cost of varint encoding has a deleterious effect on someone's workload. Should we be thinking about feature flagging this? Definitely not the kind of thing you can check the flag for on every encode/decode, but possibly the kind of thing that you could check for once at process start.

File: src/repr/src/row.rs [hunk 2]
```diff
@@ -461,6 +495,25 @@ fn read_byte(data: &[u8], offset: &mut usize) -> u8 {
...
+) -> [u8; N] {
+    debug_assert!(length <= N);
+    let mut raw = [0; N];
+    if length > 0 {
+        std::io::Write::write(&mut &mut raw[0..length], &data[*offset..*offset + length]).unwrap();
+        *offset += length;
+        // This is little-endian, so the sign bit is in the last
+        // element.
+        let is_negative = (raw[length - 1] & 0x80) != 0;
+        let fill_val = if is_negative { 255 } else { 0 };
```
Comment (mjibson) [inline src/repr/src/row.rs:511]: Why do negatives need to fill with 255?

File: src/repr/src/row.rs [hunk 3]
```diff
@@ -1982,6 +2191,15 @@ mod tests {
                 ))
                 .unwrap(),
             ),
+            Datum::Timestamp(
```
Comment (mjibson) [inline src/repr/src/row.rs:2194]: This function should also call `ScalarType::enumerate()` then `interesting_datums()` on all things returned there. Might be good to even remove all these datums, and only use that list. If there's things here we want to test, they should be added to interesting_datums.

File: src/repr/src/row.rs (hunk 2 above)
Comment (umanwizard) [reply src/repr/src/row.rs:513]: That's just how two's complement integers work. For example, -96 is `0xA0` as an i8 and `0xFFFFFFFFFFFFFFA0` as an i64.

File: src/repr/src/row.rs [hunk 4]
```diff
@@ -461,6 +495,25 @@ fn read_byte(data: &[u8], offset: &mut usize) -> u8 {
     byte
 }
 
+fn read_byte_array_sign_extending<const N: usize>(
+    data: &[u8],
+    offset: &mut usize,
+    length: usize,
+) -> [u8; N] {
+    debug_assert!(length <= N);
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply src/repr/src/row.rs:505]: I would rather not take the extra cost in this very hot code path.

File: src/repr/src/row.rs [hunk 5]
```diff
@@ -737,6 +832,78 @@ where
...
+
+    // To fit in n bytes, we require that
+    // everything but the leading sign bits fits in n*8 - 1
+    // bits. Why the -1? Because we still need one sign bit
+    // in the encoded data.
+    //
+    // We can improve this, at the cost of doubling the number of extra tags,
+    // by encoding whether the number is positive or negative in the tag.
+    // We could then replace all instances of `7F` with `FF` below.
+    if i == 0 || i.is_positive() {
```
Comment (ParkMyCar) [inline src/repr/src/row.rs:868]: IIUC we're trying to determine how many bytes are needed to encode a value? Which I think can be done more succinctly with:
```
let extra_bytes = i.leading_zeros().saturating_sub(1) / 8;
let needed_bytes = std::mem::size_of::<i64>() - extra_bytes;
```

Comment (umanwizard) [reply src/repr/src/row.rs:868]: That's actually what I initially had, check out the commit "Change how we calculate width" which switched to the nested if statement. (Except, using `leading_ones` in the case of negatives, of course).

I changed it because I suspected the count leading zeros instruction is a bit expensive on x86.

But that might be a premature micro-optimization; maybe I should change it back... it doesn't seem to have made any difference in feature benchmarks.

Comment (ParkMyCar) [reply src/repr/src/row.rs:868]: > That's actually what I initially had, check out the commit "Change how we calculate width" which switched to the nested if statement. (Except, using leading_ones in the case of negatives, of course).

Nice! 

_nerd snipe incoming:_ According to these [tables](https://www.agner.org/optimize/instruction_tables.pdf) the instruction used for counting leading zeros, `LZCNT`(which is what Rust/LLVM should use), is no more expensive then a `JMP`, so I don't think counting leading zeros should be any worse. There's other things to consider like speculative execution and binary size, which should also favor counting leading zeros, but this is 1,000% a micro optimization, so I'd go with whatever you think is most clear 😊

**Comment by benesch** [comment]
> I think the risk of a show-stopping CPU regression is not serious enough to warrant going to all that trouble. What do you think?

I have no idea! I was hoping someone with more intuition about this would have a take.

**Comment by umanwizard** [comment]
All we're doing is adding a stack of if statements when encoding signed ints. It seems hard to believe this would make a meaningful difference given how much interpreter overhead we already have in Materialize.

Maybe if you made a microbenchmark where you encoded and decoded rows with a lot of int64s over and over it would make a detectable difference, but even then IDK.

On the other hand, I did see some past runs where some kafka upsert feature benchmark was regressing. @philip-stoev do you know if that is noise that can be ignored, or should we be concerned by it?

**Review by teskje — COMMENTED** [review]
(I didn't get very far through the review, sorry!)

Did you run this change against benchmarks to see how it compares with main? I think both TPC-H and LDBC runs should be producible today with reasonable effort. I would be really interested to see what the effect on memory and CPU is.

If the benchmarks look good, I think we can live without a feature flag. I would be very hesitant to deploy such a change to prod without benchmarks though.

File: src/repr/src/row.rs (hunk 5 above)
Comment (umanwizard) [reply src/repr/src/row.rs:868]: I agree. I will change it back.

**Comment by umanwizard** [comment]
My intuition that this has no measurable perf impact seems to have been wrong. It causes the time taken on TPCH queries to increase by between 1 and 12 percent in my experiment. https://docs.google.com/spreadsheets/d/1LQ6EuU6CihlLhOYKx6lzUn87COvbUa5Pv0NRoXqA088

**Comment by benesch** [comment]
How much does it improve memory usage?

**Comment by umanwizard** [comment]
Around 10%. Not as much as I had hoped, but then again, most TPCH indexes aren't incredibly wide.

**Comment by umanwizard** [comment]
I found an optimization that makes the CPU usage closer.

The idea is to have as few if statements as possible. `min_bytes_sized` uses `lzcnt` rather than a stack of ifs. We also avoid using any ifs when computing which tag to use, and instead just do e.g. `u8::from(Tag::Int64_8) + mbs - 1`. This requires that all the tags for a given scalar type be in increasing order of actual width.

Results here: https://docs.google.com/spreadsheets/d/1pi77Gz35jfz4jt0It2ydnNl3XFZVhCSw1Vc1R6lr-N8

**Comment by umanwizard** [comment]
We collectively decided in the compute team not to land this without a feature flag. Marking as draft until that change is ready.

**Comment by umanwizard** [comment]
This is ready for review.

**Review by mjibson — COMMENTED** [review]
New coord parts lgtm.

File: src/compute-client/src/protocol/command.rs [hunk 6]
```diff
@@ -73,7 +73,10 @@ pub enum ComputeCommand<T = mz_repr::Timestamp> {
     /// dataflows according to the given [`LoggingConfig`].
     ///
     /// [Creation Stage]: super#creation-stage
-    CreateInstance(LoggingConfig),
+    CreateInstance {
+        logging_config: LoggingConfig,
+        variable_length_row_encoding: bool,
+    },
```
Comment (teskje) [inline src/compute-client/src/protocol/command.rs:79]: How about we introduce a new type `InstanceConfig` to wrap these two fields and then have a `CreateInstance(InstanceConfig)` command? I think that would simplify that replica code a bit, particularly the "has the instance config changed?" part, and even more so if we add more config options in the future.

File: src/sql/src/session/vars.rs [hunk 7]
```diff
@@ -2415,6 +2423,9 @@ impl SystemVars {
                 .lock()
                 .expect("lock poisoned")
                 .limit = u64::cast_from(*self.expect_value(&MAX_CONNECTIONS));
+        } else if name == VARIABLE_LENGTH_ROW_ENCODING.name {
+            let value = *self.expect_value(&VARIABLE_LENGTH_ROW_ENCODING);
+            mz_repr::VARIABLE_LENGTH_ROW_ENCODING.store(value, atomic::Ordering::SeqCst);
         }
```
Comment (teskje) [inline src/sql/src/session/vars.rs:2429]: Is it true that we will apply this change every time LD pushes a variable update? So when the value in LD changes while envd is running that would change the way row encoding is performed in envd? Is that fine?

File: src/expr/src/scalar/func/impls/datum.rs [hunk 8]
```diff
@@ -38,7 +38,10 @@ sqlfunc!(
         match a {
             Datum::Null => Ok(None),
             datum => {
-                let sz = mz_repr::datum_size(&datum);
+                // This will undercount sizes for installations that aren't in the
+                // `variable_length_row_encoding` treatment, but
+                // functions need to be determinstic, so there's no better we can do.
+                let sz = mz_repr::datum_size_deterministic::<true>(&datum);
```
Comment (teskje) [inline src/expr/src/scalar/func/impls/datum.rs:44]: Why do we decide to undercount rather than overcount? That latter seems safer to me, but that depends on how these SQL functions are used.

Maybe we should overcount as long as the default for the variable length encoding flag is still `false` and switch once we switch the flag default?

File: src/sql/src/session/vars.rs [hunk 9]
```diff
@@ -2415,6 +2423,9 @@ impl SystemVars {
                 .lock()
                 .expect("lock poisoned")
                 .limit = u64::cast_from(*self.expect_value(&MAX_CONNECTIONS));
+        } else if name == VARIABLE_LENGTH_ROW_ENCODING.name {
+            let value = *self.expect_value(&VARIABLE_LENGTH_ROW_ENCODING);
+            mz_repr::VARIABLE_LENGTH_ROW_ENCODING.store(value, atomic::Ordering::SeqCst);
```
Comment (teskje) [inline src/sql/src/session/vars.rs:2428]: What is the rationale for the `Ordering::SeqCst` here? I can't really say I grok the memory orderings, but I know that `SeqCst` is the strictest one and therefore potentially the most expensive one too. Would a less strict ordering suffice here?

**Review by teskje — COMMENTED** [review]
Thanks, this looks good! I have a couple questions left before I'll hit approve, the most important one being around the feature flag propagation within envd.

**Review by antiguru — COMMENTED** [review]
This seems great, but I'd feel more comfortable with:
* A repeatable benchmark that shows the performance and memory impact of this work.
* A benchmark of the impact of the feature flag.

File: src/sql/src/session/vars.rs (hunk 7 above)
Comment (umanwizard) [reply src/sql/src/session/vars.rs:2429]: I can't immediately see how it would cause problems, but for sanity, it's probably better if we always keep envd in sync with clusterd. Good catch.

File: src/sql/src/session/vars.rs (hunk 9 above)
Comment (umanwizard) [reply src/sql/src/session/vars.rs:2428]: A less strict ordering is probably fine, however, I chose to go with a stricter one to avoid having to think about the consequences of that. In general, I always choose the strictest ordering unless I have good reason to believe it's a performance bottleneck.

In this case, having the strictest possible ordering shouldn't change performance, because:

(1) on x86, reads are always atomic, so this is a no-op on the load side (it will emit the same instructions as any memory load), and
(2) although it might be slower on the store side, that is okay, since we are only storing into this value on LD changes.

Comment (teskje) [reply src/sql/src/session/vars.rs:2428]: Cool, if it doesn't matter for read performance, I agree with your reasoning!

File: src/expr/src/scalar/func/impls/datum.rs (hunk 8 above)
Comment (umanwizard) [reply src/expr/src/scalar/func/impls/datum.rs:44]: Btw, if you feel strongly that it's important, I don't mind switching this to the `false` variant

**Comment by umanwizard** [comment]
@teskje , the latest version applies your suggestions.

@antiguru: I think this is low-risk enough to land without a benchmark (assuming the flag is off) and then test the effect of the flag in prod. I tried for a while to get a good benchmark that would show the impact, but I wasn't able to find a stable enough one.

My confidence is increased by the fact that in this latest version of the code, I'm not seeing regressions in the feature benchmark as I was before.

If folks disagree about the level of risk, I think we'll need to work with QA to build proper row/decoding benchmarks before landing this.

**Comment by teskje** [comment]
It should be relatively simple to run TPC-H on staging using the loadgen, no? I think just running the unindexed queries a couple times each and then comparing the minimum times should make it pretty obvious if the feature flag introduces any noticeable regressions.

**Comment by umanwizard** [comment]
> It should be relatively simple to run TPC-H on staging using the loadgen, no

Yes, it's simple to do that, but I ran the queries hundreds of times and they did not seem to be converging to meaningful results, because the variance was too high. I suspect it also depends on factors outside our control, e.g. what machine the VM is running on (noisy neighbors, etc.).

It's a good idea to look at min instead of avg; I'll try that.

**Comment by umanwizard** [comment]
ok, I've conducted the benchmark described here: https://www.notion.so/materialize/Simple-TPC-H-benchmark-b6989dd94fff48cb9ea036724c5f9476/

My results are recorded here: 
[experiment.ods](https://github.com/MaterializeInc/materialize/files/12479526/experiment.ods)


The tl;dr: after 100 trials, all of the query timing values are within 10%; most are much closer than that (and often the new code is better, which should be impossible). I still believe that any difference is due to noise, but I don't know enough about statistical hypothesis testing to prove that claim. I haven't tested with the flag on, as that will be much easier to do in prod.

To be clear, even if I were wrong about it being noise, and there were a solid, reproducible, 10% regression, I would still be strongly in favor of landing this change so we can start testing the impact of toggling the flag in prod.

**Comment by nrainer-materialize** [comment]
I had triggered a [coverage build.](https://buildkite.com/materialize/coverage/builds/198) The coverage of new and changed code looks good to me. 👍

**Comment by teskje** [comment]
Copying the minimum times here for convenience:

```
query  this PR [s]  main [s]   diff [%]
---------------------------------------
q01     935.587      916.974   +2.0
q02     196.989      195.662   +0.7
q02     237.693      233.465   +1.8
q04     311.417      305.489   +1.9
q05      91.486       89.301   +2.4
q06     193.859      184.580   +4.8
q07     162.440      164.100   -1.0
q08     162.187      162.454   -0.2
q09     287.350      284.728   +0.9
q10    1133.722     1100.447   +2.9
q11     321.310      315.712   +1.7
q12     188.324      185.523   +1.5
q13     306.387      301.118   +1.7
q14     200.039      188.660   +5.7
q15     183.428      174.231   +5.0
q16     293.211      287.440   +2.0
q17     202.975      192.530   +5.1
q18    1329.595     1304.241   +1.9
q19     237.867      228.036   +4.1
q20    8191.298     8084.327   +1.3
q21     266.459      261.695   +1.8
q22     294.945      290.933   +1.4
```

I also won't pretend to know anything about statistics, so I can't say for sure that this isn't all just noise. But it _looks_ like a performance regression is visible, at least with the queries that have a ~5% diff.

But I'm fine with merging this, given that our troubles usually are with memory usage, not response times.

**Comment by umanwizard** [comment]
Slight correction, @teskje : those are milliseconds, not seconds.

**Comment by teskje** [comment]
🤦 Thanks, fixed!

File: src/compute/src/server.rs [hunk 10]
```diff
@@ -401,7 +401,7 @@ impl<'w, A: Allocate + 'static> Worker<'w, A> {
 
     fn handle_command(&mut self, response_tx: &mut ResponseSender, cmd: ComputeCommand) {
         match &cmd {
-            ComputeCommand::CreateInstance(_) => {
+            ComputeCommand::CreateInstance { .. } => {
```
Comment (antiguru) [inline src/compute/src/server.rs:404]: Same here.

File: src/sql/src/session/vars.rs [hunk 11]
```diff
@@ -849,6 +849,13 @@ static DEFAULT_WEBHOOKS_SECRETS_CACHING_TTL_SECS: Lazy<usize> = Lazy::new(|| {
     usize::cast_from(mz_secrets::cache::DEFAULT_TTL_SECS.load(std::sync::atomic::Ordering::Relaxed))
 });
 
+const VARIABLE_LENGTH_ROW_ENCODING: ServerVar<bool> = ServerVar {
+    name: UncasedStr::new("variable_length_row_encoding"),
+    value: &false,
+    description: "Determines whether `repr::row::VARIABLE_LENGTH_ENCODING` is set. See that module for details.",
+    internal: true,
+};
+
```
Comment (antiguru) [inline src/sql/src/session/vars.rs:858]: Why not use the `feature_flags!` macro here?

File: src/testdrive/src/action/sql.rs [hunk 12]
```diff
@@ -162,7 +162,7 @@ pub async fn run_sql(mut cmd: SqlCommand, state: &mut State) -> Result<ControlFl
                         .collect::<Vec<_>>()
                         .join("\n");
 
-                    bail!("the on-disk state of the catalog does not match its in-memory state:\n{diff}");
+                    bail!("the in-memory state of the catalog does not match its on-disk state:\n{diff}");
```
Comment (antiguru) [inline src/testdrive/src/action/sql.rs:165]: Unrelated change?

Comment (umanwizard) [reply src/testdrive/src/action/sql.rs:165]: It's related in the sense that this message confused me while I was debugging test failures related to this change. The order of words in the message now matches the order of strings being diffed.

**Review by antiguru — APPROVED** [review]
